Lesson19/elifbasak/script.js

```diff
@@ -0,0 +1,163 @@
+   - Remove the item from the `#items` array if it exists.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 4: Calculate the Total Cost
+-----------------------------------------------------------
+1. Add a `getTotal` method to the `ShoppingCart` class.
+2. The method should:
+   - Calculate and return the total cost of all items in 
+     the cart.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 5: Apply a Discount
+-----------------------------------------------------------
+1. Add an `applyDiscount` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept a discount code (e.g., 'SAVE10', 'SAVE20').
+   - Apply a percentage discount to the total cost if the 
+     code is valid.
+3. Use an object to store discount codes and their values.
+*/
+
+class ShoppingCart {
+  #items;
+  #discountCodes;
+
+  constructor() {
+    this.#items = [];
+    this.#discountCodes = { SAVE10: 0.1, SAVE20: 0.2, SAVE30: 0.3 }; // I added the codes as a private property on the class.
+  }
+
+  viewCart() {
+    console.log("Viewing cart....");
+    for (const item of this.#items) {
+      console.log(item);
+    }
+  }
+
+  addItem(name, price, quantity) {
+    console.log("Adding new item to the cart...");
+    for (const item of this.#items) {
+      if (name === item.name) {
+        console.log("Item found in the cart, increasing the quantity.");
+        item.quantity++;
+        return;
+      }
+    }
+
+    const id = new Date().getMilliseconds();
+    this.#items.push({ id, name, price, quantity });
+  }
+
+  // In real life, we usually use id to find elements or do any manipulations
+  removeItem(name) {
+    console.log("Removing the item...");
+    for (let i = 0; i < this.#items.length; i++) {
+      const currentElement = this.#items[i];
+      if (name.toLowerCase() === currentElement.name.toLowerCase()) {
+        this.#items.splice(i, 1);
+        return;
+      }
+    }
+  }
+
+  getTotal() {
+    console.log("Calculating the total cost of the cart....");
+    let total = 0;
+    for (const item of this.#items) {
+      total += item.quantity * item.price.amount;
+    }
+    console.log(`Total amount of the cart is ${total} TRY.`);
+    return total;
+  }
+
+  applyDiscount(discountCode) {
+    console.log("Applying discount....");
+    if (typeof discountCode !== "string") {
+      console.log("Please provide the discount code in a string format.");
+      return;
+    }
+    const discountRate = this.#discountCodes[discountCode];
+    if (this.#discountCodes[discountCode]) {     //Error validation is simplified.
+      const totalAmount = this.getTotal();
+      const discountedTotal = totalAmount - totalAmount * discountRate;
+      console.log(`After the ${discountRate * 100}% discount, the new total is ${discountedTotal}`);
+    return discountedTotal;
+    } else {
+      console.log("The discount code is not valid.");
+      return this.getTotal();
+    }
+  }
```

Comment:
After handling the error validation above, this whole block can simply become the `else` statement. Now `applyDiscount` will first check `if` the code does not match any key in the `discountCodes` object and return the total value. If it matches, the `else` part will calculate the discount and return the discounted value.

```suggestion
else {
    const discountRate = this.#discountCodes[discountCode];
      const totalAmount = this.getTotal();
      const discountedTotal = totalAmount - totalAmount * discountRate;
      console.log(`After the ${discountRate * 100}% discount, the new total is ${discountedTotal}`);
    return discountedTotal;
    }
  }
```